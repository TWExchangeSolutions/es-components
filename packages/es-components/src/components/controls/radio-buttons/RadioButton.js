import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider, withTheme } from 'styled-components';
import { noop } from 'lodash';
import viaTheme from 'es-components-via-theme';

import Label from '../Label';
import getRadioFillVariables from './radio-fill-variables';

function radioFill(color) {
  return `
    background-color: ${color};
    border-radius: 100%;
    content: '';
    display: block;
    height: 13px;
    left: 3px;
    position: absolute;
    width: 13px;
    top: 3px;
    transition: background 0.25s linear;
  `;
}

const RadioLabel = styled(Label)`
  color: ${props => (props.error ? props.theme.colors.danger : 'inherit')};
  display: ${props => (props.inline ? 'inline-flex' : 'inherit')};
  font-size: ${props => props.theme.sizes.baseFontSize}px;
  font-weight: normal;
  margin-right: ${props => (props.inline ? '20px' : 'initial')};
  min-height: 25px;
  position: relative;
  padding: 5px 0;
  text-transform: none;

  &:hover .radio-fill:before {
    ${props => radioFill(props.hoverFillColor)};
  }
`;

const RadioInput = styled.input`
  opacity: 0;
  position: absolute;

  &:focus ~ .radio-fill {
    box-shadow: 0 0 3px 3px ${props => props.theme.colors.inputFocus};
  }
`;

const RadioText = styled.span`
  align-self: center;
  margin-left: 30px;
`;

const RadioDisplay = styled.span`
  border: 3px solid ${props => props.borderColor};
  border-radius: 100%;
  box-sizing: border-box;
  display: block;
  height: 25px;
  position: absolute;
  width: 25px;

  &:before {
    ${props => radioFill(props.fill)};
  }
`;

export function RadioButton({
  optionText,
  name,
  checked = false,
  id,
  isDisabled = false,
  inline = true,
  onClick = noop,
  value,
  hasError = false,
  theme,
  ariaHide,
  ...radioProps
}) {
  const { hover, fill } = getRadioFillVariables(
    checked,
    isDisabled,
    hasError,
    theme.colors
  );
  const radioDisplayFill = checked ? fill : theme.colors.white;

  const labelProps = {
    checked,
    inline,
    disabled: isDisabled,
    htmlFor: id,
    hoverFillColor: hover,
    error: hasError
  };

  return (
    <ThemeProvider theme={theme}>
      <RadioLabel {...labelProps}>
        <RadioInput
          type="radio"
          name={name}
          id={id}
          onClick={onClick}
          value={value}
          disabled={isDisabled}
          {...radioProps}
        />
        <RadioDisplay
          className="radio-fill"
          borderColor={fill}
          fill={radioDisplayFill}
        />
        <RadioText aria-hidden={ariaHide}>{optionText}</RadioText>
      </RadioLabel>
    </ThemeProvider>
  );
}

RadioButton.propTypes = {
  optionText: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  id: PropTypes.string,
  isDisabled: PropTypes.bool,
  inline: PropTypes.bool,
  onClick: PropTypes.func,
  value: PropTypes.any,
  ariaHide: PropTypes.bool,
  hasError: PropTypes.bool,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object
};

RadioButton.defaultProps = {
  theme: viaTheme
};

export default withTheme(RadioButton);
