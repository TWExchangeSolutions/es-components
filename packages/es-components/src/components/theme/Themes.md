To apply a theme you can either wrap your root react app component in a `<ThemeProvider theme='[themeName]'>` wrapper component, or provide a theme to individual components via the `theme` prop. The `ThemeProvider` wrapper provides the supplied theme to any components underneath it. See [styled-components Theming](https://www.styled-components.com/docs/advanced#theming).

Example:
```html
import { ThemeProvider } from 'styled-components';
import { viaTheme } from 'es-components';
...
render(
	<ThemeProvider theme={viaTheme}>
		<MyComponentOrApp />
	</ThemeProvider>
);
```

Three themes are provided with this library: `defaultTheme`, `wtwTheme`, and `viaTheme`.

To create your own theme, you could simply copy one of the themes (a simple javascript object), modify the values, and supply that object to the `ThemeProvider`.

### Theme colors

<table style="margin-bottom: 1.5em">
	<tbody>
		<tr><th colspan="7" style="text-align: left">Default</th></tr>
		<tr>
			<td>Primary <div style="background-color: #007BFF; padding: 1em; width: 50px"></div></td>
			<td>Default <div style="background-color: #6c757d; padding: 1em; width: 50px"></div></td>
			<td>Info <div style="background-color: #17a2b8; padding: 1em; width: 50px"></div></td>
			<td>Success <div style="background-color: #28a745; padding: 1em; width: 50px"></div></td>
			<td>Warning <div style="background-color: #ffc107; padding: 1em; width: 50px"></div></td>
			<td>Danger <div style="background-color: #dc3545; padding: 1em; width: 50px"></div></td>
		</tr>
	</tbody>
</table>

<table style="margin-bottom: 1.5em">
	<tbody>
		<tr><th colspan="7" style="text-align: left">WTW</th></tr>
		<tr>
			<td>Primary <div style="background-color: #5a0c6f; padding: 1em; width: 50px"></div></td>
			<td>Default <div style="background-color: #444; padding: 1em; width: 50px"></div></td>
			<td>Info <div style="background-color: #1b6284; padding: 1em; width: 50px"></div></td>
			<td>Success <div style="background-color: #060; padding: 1em; width: 50px"></div></td>
			<td>Warning <div style="background-color: #ebaf00; padding: 1em; width: 50px"></div></td>
			<td>Danger <div style="background-color: #af140c; padding: 1em; width: 50px"></div></td>
		</tr>
	</tbody>
</table>

<table style="margin-bottom: 1.5em">
	<tbody>
		<tr><th colspan="7" style="text-align: left">Via</th></tr>
		<tr>
			<td>Primary <div style="background-color: #007fa7; padding: 1em; width: 50px"></div></td>
			<td>Default <div style="background-color: #d8d8d8; padding: 1em; width: 50px"></div></td>
			<td>Info <div style="background-color: #069; padding: 1em; width: 50px"></div></td>
			<td>Success <div style="background-color: #298544; padding: 1em; width: 50px"></div></td>
			<td>Warning <div style="background-color: #b35f00; padding: 1em; width: 50px"></div></td>
			<td>Danger <div style="background-color: #c00; padding: 1em; width: 50px"></div></td>
		</tr>
	</tbody>
</table>

### Fonts

Apart from the icon set, this library does not have a specific typography and should work with whatever font you have defined in your project.
