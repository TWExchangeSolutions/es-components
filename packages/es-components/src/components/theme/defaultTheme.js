const white = '#fff';
const gray1 = '#f8f9fa';
const gray2 = '#e9ecef';
const gray3 = '#dee2e6';
const gray4 = '#ced4da';
const gray5 = '#adb5bd';
const gray6 = '#868e96';
// const gray7 = '#495057';
const gray8 = '#343a40';
const gray9 = '#212529';
const black = '#000';

const grayLightest = gray1;
const grayLighter = gray2;
const grayLight = gray3;
const gray = gray5;
const grayDark = gray6;
const grayDarker = gray8;
const grayDarkest = gray9;

const defaultColor = '#6c757d';
const defaultHover = '#5a6268';
const defaultBtnText = white;
const primary = '#007BFF';
const primaryHover = '#0062cc';
const info = '#17A2B8';
const infoHover = '#138496';
const success = '#28A745';
const successHover = '#218838';
const warning = '#FFC107';
const warningHover = '#e0a800';
const danger = '#DC3545';
const dangerHover = '#c82333';
const advisor = warning;
const popoverHeader = info;

const boxShadowLight = 'rgba(0, 0, 0, 0.075)';
const boxShadowDark = 'rgba(0, 0, 0, 0.5)';

const inputBorder = gray4;
const inputFocus = '#66afe9';
const inputBoxShadow = 'rgba(102, 175, 233, 0.6)';

// Datepicker colors
const navArrow = gray;
const navArrowHover = grayDark;
const selected = '#216ba5';
const hover = '#1d5d90';
const keyboard = '#2a87d0';
const inRange = 'rgba(33, 107, 165, 0.5)';
const highlight = '#3dcc4a';
const highlightHover = '#32be3f';

const theme = {
  colors: {
    black,
    grayDarkest,
    grayDarker,
    grayDark,
    gray,
    grayLight,
    grayLighter,
    grayLightest,
    white,
    defaultColor,
    defaultHover,
    defaultBtnText,
    primary,
    primaryHover,
    success,
    successHover,
    info,
    infoHover,
    warning,
    warningHover,
    danger,
    dangerHover,
    advisor,
    boxShadowLight,
    boxShadowDark,
    inputFocus,
    popoverHeader
  },
  datepickerColors: {
    navArrow,
    navArrowHover,
    selected,
    hover,
    keyboard,
    inRange,
    highlight,
    highlightHover
  },
  notificationStyles: {
    success: {
      bgColor: success,
      textColor: white
    },
    info: {
      bgColor: info,
      textColor: white
    },
    warning: {
      bgColor: warning,
      textColor: black
    },
    danger: {
      bgColor: danger,
      textColor: white
    },
    advisor: {
      bgColor: advisor,
      textColor: black
    }
  },
  buttonStyles: {
    buttonsNormal: {
      // these button types are dynamic, so you can add/remove/rename them
      primary: {
        bgColor: primary,
        textColor: white,
        hoverBgColor: primaryHover,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        boxShadowColor: primaryHover
      },
      default: {
        bgColor: defaultColor,
        textColor: white,
        hoverBgColor: defaultHover,
        hoverTextColor: white,
        activeBgColor: defaultHover,
        activeTextColor: white,
        boxShadowColor: defaultHover
      },
      success: {
        bgColor: success,
        textColor: white,
        hoverBgColor: successHover,
        hoverTextColor: white,
        activeBgColor: successHover,
        activeTextColor: white,
        boxShadowColor: successHover
      },
      info: {
        bgColor: info,
        textColor: white,
        hoverBgColor: infoHover,
        hoverTextColor: white,
        activeBgColor: infoHover,
        activeTextColor: white,
        boxShadowColor: infoHover
      },
      warning: {
        bgColor: warning,
        textColor: black,
        hoverBgColor: warningHover,
        hoverTextColor: black,
        activeBgColor: warningHover,
        activeTextColor: black,
        boxShadowColor: warningHover
      },
      danger: {
        bgColor: danger,
        textColor: white,
        hoverBgColor: dangerHover,
        hoverTextColor: white,
        activeBgColor: dangerHover,
        activeTextColor: white,
        boxShadowColor: dangerHover
      }
    },
    buttonsOutline: {
      // these can be independent of the main buttons, but generally
      // should have matching sets
      primary: {
        bgColor: white,
        textColor: primary,
        hoverBgColor: primary,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        borderColor: primary
      },
      default: {
        bgColor: white,
        textColor: defaultColor,
        hoverBgColor: defaultColor,
        hoverTextColor: white,
        activeBgColor: defaultHover,
        activeTextColor: white,
        borderColor: defaultColor
      },
      success: {
        bgColor: white,
        textColor: success,
        hoverBgColor: success,
        hoverTextColor: white,
        activeBgColor: successHover,
        activeTextColor: white,
        borderColor: success
      },
      info: {
        bgColor: white,
        textColor: info,
        hoverBgColor: info,
        hoverTextColor: white,
        activeBgColor: infoHover,
        activeTextColor: white,
        borderColor: info
      },
      warning: {
        bgColor: white,
        textColor: warning,
        hoverBgColor: warning,
        hoverTextColor: black,
        activeBgColor: warningHover,
        activeTextColor: black,
        borderColor: warning
      },
      danger: {
        bgColor: white,
        textColor: danger,
        hoverBgColor: danger,
        hoverTextColor: white,
        activeBgColor: dangerHover,
        activeTextColor: white,
        borderColor: danger
      }
    }
  },
  // buttonSizes should always have default, lg, sm, xs
  buttonSizes: {
    lg: {
      borderRadius: '3px',
      fontSize: '26px',
      padding: '8px 20px 5px'
    },
    default: {
      borderRadius: '2px',
      fontSize: '18px',
      padding: '5px 15px 4px'
    },
    sm: {
      borderRadius: '2px',
      fontSize: '14px',
      padding: '4px 10px 3px'
    },
    xs: {
      borderRadius: '2px',
      fontSize: '12px',
      padding: '3px 10px 2px'
    }
  },
  validationIconName: {
    success: 'ok',
    warning: 'warning-sign',
    danger: 'remove'
  },
  validationTextColor: {
    success,
    warning: warningHover,
    danger
  },
  validationInputColor: {
    success: {
      borderColor: success,
      boxShadow: `inset 0 0 5px ${success}`,
      focusBorderColor: success,
      focusBoxShadow: `0 0 6px ${success}`
    },
    warning: {
      borderColor: warning,
      boxShadow: `inset 0 0 5px ${warning}`,
      focusBorderColor: warning,
      focusBoxShadow: `0 0 6px ${warning}`
    },
    danger: {
      borderColor: danger,
      boxShadow: `inset 0 0 5px ${danger}`,
      focusBorderColor: danger,
      focusBoxShadow: `0 0 6px ${danger}`
    },
    default: {
      borderColor: inputBorder,
      boxShadow: `inset 0 1px 1px ${boxShadowLight}`,
      focusBorderColor: inputFocus,
      focusBoxShadow: `0 0 8px ${inputBoxShadow}`
    }
  },
  sizes: {
    baseFontSize: 18,
    baseLineHeight: 24
  },
  screenSize: {
    retina: '320px',
    phone: '480px',
    tablet: '768px',
    desktop: '992px',
    widescreen: '1200px'
  }
};

export default theme;
