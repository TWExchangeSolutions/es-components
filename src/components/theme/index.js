export const colors = {
  white: '#fff',
  red: '#af140c',
  black: '#000',
  grayDarkest: '#444',
  grayDark: '#888',
  gray: '#aaa',
  grayLight: '#ddd',
  grayLighter: '#eee',
  success: '#060',
  accent: '#5a0c6f',
  information: '#1b6284',
  warning: '#ebaf00',
  danger: '#af140c',
  advisor: '#ff6310',
  errorText: '#df190f'
};

export const sizes = {
  baseFontSize: 18,
  baseLightHeight: 24
};
