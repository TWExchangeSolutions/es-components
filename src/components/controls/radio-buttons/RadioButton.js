import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { noop } from 'lodash';

import Label from '../Label';
import { colors, sizes } from '../../theme';

import getRadioFillVariables from './radio-fill-variables';

function radioFill(color) {
  return `
    background-color: ${color};
    border-radius: 100%;
    content: '';
    display: block;
    height: 13px;
    left: 3px;
    position: absolute;
    width: 13px;
    top: 3px;
    transition: background 0.25s linear;
  `;
}

const RadioLabel = styled(Label)`
  color: ${props => (props.error ? colors.danger : 'inherit')};
  display: ${props => (props.inline ? 'inline-flex' : 'inherit')};
  font-size: ${sizes.baseFontSize}px;
  font-weight: normal;
  margin-right: ${props => (props.inline ? '20px' : 'initial')};
  min-height: 25px;
  position: relative;
  padding: 5px 0;
  text-transform: none;

  &:hover .radio-fill:before { ${props => radioFill(props.hoverFillColor)} }
`;

const RadioInput = styled.input`
  opacity: 0;
  position: absolute;

  &:focus ~ .radio-fill {
    box-shadow: 0 0 3px 3px #83bffc
  }
`;

const RadioText = styled.span`
  align-self: center;
  margin-left: 30px;
`;

const RadioDisplay = styled.span`
  border: 3px solid ${props => props.borderColor};
  border-radius: 100%;
  box-sizing: border-box;
  display: block;
  height: 25px;
  position: absolute;
  width: 25px;

  &:before { ${props => radioFill(props.fill)} }
`;

function RadioButton({
  optionText,
  name,
  checked = false,
  id,
  disabled = false,
  inline = true,
  onChange = noop,
  value,
  isInErrorState = false,
  ...radioProps
}) {
  const { hover, fill } = getRadioFillVariables(
    checked,
    disabled,
    isInErrorState
  );
  const radioDisplayFill = checked ? fill : colors.white;

  const labelProps = {
    checked,
    inline,
    disabled,
    htmlFor: id,
    hoverFillColor: hover,
    error: isInErrorState
  };

  return (
    <RadioLabel {...labelProps}>
      <RadioInput
        type="radio"
        name={name}
        id={id}
        onChange={onChange}
        value={value}
        disabled={disabled}
        {...radioProps}
      />
      <RadioDisplay
        className="radio-fill"
        borderColor={fill}
        fill={radioDisplayFill}
      />
      <RadioText>{optionText}</RadioText>
    </RadioLabel>
  );
}

RadioButton.propTypes = {
  optionText: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  id: PropTypes.string,
  disabled: PropTypes.bool,
  inline: PropTypes.bool,
  onChange: PropTypes.func,
  value: PropTypes.any,
  isInErrorState: PropTypes.bool
};

export default RadioButton;
