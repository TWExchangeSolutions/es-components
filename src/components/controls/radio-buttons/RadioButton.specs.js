/* eslint-env jest */
/* eslint-disable max-len */

import React from 'react';
import renderer from 'react-test-renderer';

import { colors } from '../../theme';

import getRadioFillVariables from './radio-fill-variables';
import RadioButton from './RadioButton';

describe('RadioButton component', () => {
  it('renders as expected', () => {
    const tree = renderer.create(<RadioButton id="test" optionText="test" name="rad" />);

    expect(tree).toMatchSnapshot();
  });
});

describe('getRadioFillVariables', () => {
  it('returns { fill: gray, hover: white } when unchecked, disabled, and valid', () => {
    const checked = false;
    const disabled = true;
    const invalid = false;

    const { fill, hover } = getRadioFillVariables(checked, disabled, invalid);

    expect(fill).toBe(colors.gray);
    expect(hover).toBe(colors.white);
  });

  it('returns { fill: accent } when checked, not disabled, and valid', () => {
    const checked = true;
    const disabled = false;
    const invalid = false;

    const { fill } = getRadioFillVariables(checked, disabled, invalid);

    expect(fill).toBe(colors.accent);
  });

  it('returns { fill: gray } when checked, disabled, and valid', () => {
    const checked = true;
    const disabled = true;
    const invalid = false;

    const { fill } = getRadioFillVariables(checked, disabled, invalid);

    expect(fill).toBe(colors.gray);
  });

  it('returns { fill: danger, hover: danger } when unchecked, not disabled, and invalid', () => {
    const checked = false;
    const disabled = false;
    const invalid = true;

    const { fill, hover } = getRadioFillVariables(checked, disabled, invalid);

    expect(fill).toBe(colors.danger);
    expect(hover).toBe(colors.danger);
  });

  it('returns { fill: danger } when checked, not disabled, and invalid', () => {
    const checked = true;
    const disabled = false;
    const invalid = true;

    const { fill } = getRadioFillVariables(checked, disabled, invalid);

    expect(fill).toBe(colors.danger);
  });

  it('returns { fill: grayDarker, hover: grayDarker } when unchecked, not disabled, and valid', () => {
    const checked = false;
    const disabled = false;
    const invalid = false;

    const { fill, hover } = getRadioFillVariables(checked, disabled, invalid);

    expect(fill).toBe(colors.grayDarker);
    expect(hover).toBe(colors.grayDarker);
  });
});
