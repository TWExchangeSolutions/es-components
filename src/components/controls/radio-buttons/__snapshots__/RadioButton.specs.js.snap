exports[`RadioButton component renders as expected 1`] = `
<label
  checked={false}
  className="cvaaHl"
  disabled={false}
  htmlFor="test">
  <input
    className="djIxZF"
    disabled={false}
    id="test"
    name="rad"
    onChange={[Function]}
    type="radio"
    value={undefined} />
  <span
    className="radio-fill fDIfVH"
    fill="#fff" />
  <span
    className="cxSMDT">
    test
  </span>
</label>
`;
